### PR TITLE
8288984: Simplification in java.lang.Runtime::exit

### PR DESCRIPTION
**Pull request by rjernst**
This is a followup to simplify Shutdown.exit after the removal of
finalizers (https://bugs.openjdk.org/browse/JDK-8198250). Once agreement
on the approach has been reached in this PR, a CSR will be filed to
propose the spec change to Runtime.exit.

---------
### Progress
- [x] Change must be properly reviewed (1 review required, with at least 1 [Reviewer](https://openjdk.org/bylaws#reviewer))
- [x] Change must not contain extraneous whitespace
- [x] Commit message must refer to an issue
- [x] Change requires a CSR request to be approved

### Issues
 * [JDK-8288984](https://bugs.openjdk.org/browse/JDK-8288984): Simplification in java.lang.Runtime::exit
 * [JDK-8289824](https://bugs.openjdk.org/browse/JDK-8289824): Simplification in java.lang.Runtime::exit (**CSR**)

### Reviewers
 * [David Holmes](https://openjdk.org/census#dholmes) (@dholmes-ora - **Reviewer**)
 * [Chris Hegarty](https://openjdk.org/census#chegar) (@ChrisHegarty - **Reviewer**)
 * [Alan Bateman](https://openjdk.org/census#alanb) (@AlanBateman - **Reviewer**)
 * [Kim Barrett](https://openjdk.org/census#kbarrett) (@kimbarrett - **Reviewer**)

### Reviewing
<details><summary>Using <code>git</code></summary>

Checkout this PR locally: \
`$ git fetch https://git.openjdk.org/jdk pull/9351/head:pull/9351` \
`$ git checkout pull/9351`

Update a local copy of the PR: \
`$ git checkout pull/9351` \
`$ git pull https://git.openjdk.org/jdk pull/9351/head`

</details>
<details><summary>Using Skara CLI tools</summary>

Checkout this PR locally: \
`$ git pr checkout 9351`

View PR using the GUI difftool: \
`$ git pr show -t 9351`

</details>
<details><summary>Using diff file</summary>

Download this PR as a diff file: \
<a href="https://git.openjdk.org/jdk/pull/9351.diff">https://git.openjdk.org/jdk/pull/9351.diff</a>

</details>

### CODE REVIEW

**Comment by dholmes-ora** [comment]
There is not supposed to be any spec change in relation to this issue. The behaviour should be as specified; it was the coding that was intended to be simplified.

File: src/java.base/share/classes/java/lang/Runtime.java [hunk 1]
```diff
@@ -83,10 +83,9 @@ private Runtime() {}
      * until they finish.  Once this is done the virtual machine
      * {@linkplain #halt halts}.
      *
-     * <p> If this method is invoked after all shutdown hooks have already
-     * been run and the status is nonzero then this method halts the
-     * virtual machine with the given status code. Otherwise, this method
-     * blocks indefinitely.
+     * <p> Invocations of this method block indefinitely. It is therefore
+     * inadvisable to invoke this method from a shutdown hook as it will
+     * cause deadlock.
```
Comment (dholmes-ora) [inline src/java.base/share/classes/java/lang/Runtime.java:88]: This is inaccurate. The method only blocks indefinitely when shutdown has already commenced.

**Comment by dholmes-ora** [comment]
Sorry, I did not think this issue was intended to change the specification in any way, but I see now that it actually does - whether that was the intent or not.

File: src/java.base/share/classes/java/lang/Runtime.java (hunk 1 above)
Comment (rjernst) [reply src/java.base/share/classes/java/lang/Runtime.java:88]: If a shutdown hook is running, then shutdown has started, right? This new wording isn’t really a change, it’s the current behavior (the “otherwise” portion). But the wording is meant to be more accurate for the case of running exit from shutdown hooks. The change is the removal of the first case, which is what the code portion of the PR removes.

Comment (AlanBateman) [reply src/java.base/share/classes/java/lang/Runtime.java:88]: The first paragraph of the method description already makes it clear that the method never returns normally. One option for the third paragraph is to just remove it, another is to replace it with wording that specifies that the first thread to call exit is the winner and the exit code provided by other threads that attempt to exit around the same time (or while exit hooks execute) will be ignored. I think there is merit with adding a warning that a shutdown hook invoking exit will deadlock.

**Review by kristylee88 — APPROVED** [review]

File: src/java.base/share/classes/java/lang/Runtime.java (hunk 1 above)
Comment (rjernst) [reply src/java.base/share/classes/java/lang/Runtime.java]: Thanks for the feedback, they are all good points. I opted to make the behavior with multiple threads more clear, see [8e7d527](https://github.com/openjdk/jdk/pull/9351/commits/8e7d527a182933818bd2d7f0f1b799dac52663be).

**Comment by ChrisHegarty** [comment]
> Sorry, I did not think this issue was intended to change the specification in any way, but I see now that it actually does - whether that was the intent or not.

While, maybe, not strictly envisioned by the JIRA issue, the specification clarification is IMO a good addition. We should update the synopsis to make it clear that the simplification is to j.l.Runtime::exit.

**Review by kristylee88 — APPROVED** [review]

File: src/java.base/share/classes/java/lang/Runtime.java [hunk 2]
```diff
@@ -83,10 +83,10 @@ private Runtime() {}
...
      * {@linkplain #halt halts}.
      *
-     * <p> If this method is invoked after all shutdown hooks have already
-     * been run and the status is nonzero then this method halts the
-     * virtual machine with the given status code. Otherwise, this method
-     * blocks indefinitely.
+     * <p> If this method is invoked from more than one thread the status code
+     * of the first invocation will be used; the status codes from other invocations
+     * will be ignored. If this method is invoked from a shutdown hook the system
+     * will deadlock.
```
Comment (dholmes-ora) [inline src/java.base/share/classes/java/lang/Runtime.java:89]: Expressing this accurately is tricky - what is "first" here? I suggest the following:

> Invocations of this method are serialized such that only one invocation will actually proceed with the shutdown sequence and terminate the VM with the given status code. All other invocations will block indefinitely. If this method is invoked from a shutdown hook the system will deadlock.

**Review by dholmes-ora — CHANGES_REQUESTED** [review]

File: src/java.base/share/classes/java/lang/Runtime.java (hunk 2 above)
[the remark this answers is not in the document]
Comment (ChrisHegarty) [reply src/java.base/share/classes/java/lang/Runtime.java:89]: > Is "deadlock" accurate?
Yes.

In the context of the specification, "shutdown hook" means _application_ shutdown hook - as far as the specification is concerned, application shutdown hooks are the only kind of hooks. Right?

For example, the following will deadlock (when run with the changes in this PR):

```
public class TestHook {
  public static void main(String... arg) {
    Thread hook = new Thread("my-hook") {
      @Override
      public void run() {
        System.exit(1);
      }
    };
    Runtime.getRuntime().addShutdownHook(hook);
    System.exit(0);
  }
}
```

[the remark this answers is not in the document]
Comment (dholmes-ora) [reply src/java.base/share/classes/java/lang/Runtime.java:89]: It is a general deadlock, not a monitor based deadlock: the thread that called exit and holds the lock has to join() the hook thread. The hook thread blocks on the lock held by the exiting thread. Neither thread can progress.

**Comment by dholmes-ora** [comment]
> Let's say we've run shutdown so run all the hooks but not halted. Then someone calls exit(0). That seems to suggest the call will block indefinitely, which is neither desirable nor what was actually implemented.

If the hook threads do not halt then the exiting thread (which holds the lock) blocks forever in the join(). Any other call to exit blocks trying to acquire the lock. That has always been the way this works - if hook threads don't terminate then the VM doesn't (unless someone calls halt() directly). That is one of the things the window you suggested be closed, allowed - a call to exit(non-zero) could force a call to halt().

File: src/java.base/share/classes/java/lang/Runtime.java (hunk 2 above)
Comment (rjernst) [reply src/java.base/share/classes/java/lang/Runtime.java]: I reworded with the suggested text in mind. I also added another sentence referencing native signal handlers, which may also initiate shutdown. I think this clarification is important so that it is clear the status code of all invocations from Java may still be ignored. See [2253259](https://github.com/openjdk/jdk/pull/9351/commits/2253259c82b13e7b2186429a80cc42497235b035).

Comment (AlanBateman) [reply src/java.base/share/classes/java/lang/Runtime.java]: I think the wording in the latest commit (9d972b) is problematic. One reason is that you've changed it to "will run shutdown hooks" but it doesn't run the hooks, it starts them, and so conflicts with the previous paragraph. I also think "That invocation may be initiated via platform specific signal handlers" is confusing here as there is no notion of "signal handler" to reference. There may be scope for text elsewhere, maybe with an implNote for signals such as HUP, but I don't think this is the PR for this. So I think it better to try the wording that David suggested and see if it needs any improvement.

Comment (ChrisHegarty) [reply src/java.base/share/classes/java/lang/Runtime.java]: YAO (Yet Another Opinion)!  But I think we're actually converging.

David's wording:
>Invocations of this method are serialized such that only one invocation will actually proceed with the shutdown sequence and terminate the VM with the given status code.

... gives the impression that an invocation of this method WILL terminate the VM with the given status code - which is not actually true, given the potential for signals. This is already alluded to in `Runtime::addShutdownHook`. I agree that this could be resolved separately, but we should _allow_ for it.
 
The original wording:
> If this method is invoked after all shutdown hooks have already been run ... 

... seems quite clever (and allows for signals). Maybe we could:
> If this method is invoked after shutdown hooks have already been started it will block indefinitely. If this method is invoked from a shutdown hook the system will deadlock.

File: src/java.base/share/classes/java/lang/Runtime.java [hunk 3]
```diff
@@ -83,10 +83,12 @@ private Runtime() {}
      * until they finish.  Once this is done the virtual machine
      * {@linkplain #halt halts}.
      *
-     * <p> If this method is invoked from more than one thread the status code
-     * of the first invocation will be used; the status codes from other invocations
-     * will be ignored. If this method is invoked from a shutdown hook the system
-     * will deadlock.
+     * <p> Shutdown is serialized such that only one invocation will run
+     * shutdown hooks and terminate the VM with the given status code. That
+     * invocation may be initiated via platform specific signal handlers. All
```
Comment (dholmes-ora) [inline src/java.base/share/classes/java/lang/Runtime.java:88]: Why are we mentioning signal handlers here? How is that relevant?

Comment (rjernst) [reply src/java.base/share/classes/java/lang/Runtime.java:88]: Signal handlers for eg SIGTERM invoke Shutdown.shutdown. That method holds the same lock as Shutdown.exit and runs shutdown hooks. So if a signal handler triggers shutdown, and before the system halts Runtime.exit is invoked, the status passed to Runtime.exit will be ignored.

Comment (dholmes-ora) [reply src/java.base/share/classes/java/lang/Runtime.java:88]: First, the signal handlers actually invoke Shutdown.exit, not Shutdown.shutdown  - see Terminator.setup(). (If they did the latter, like DestroyJavaVM, then an exit(status) call from another thread could still do the final VM halt(status)

But now you are opening a can of worms. There are multiple ways for the VM to initiate termination - are you going to try and describe here how all of them potentially interact?  

What you are really stating here is that other parts of the JDK can invoke System.exit, but that is for them to specify where such things are specified, it isn't for exit() to try and list them all. All we have to do here is describe how multiple calls to exit() behave.

File: src/java.base/share/classes/java/lang/Runtime.java [hunk 4]
```diff
@@ -83,10 +83,12 @@ private Runtime() {}
...
+     * other invocations will block indefinitely, and their supplied exit
+     * statuses will be ignored. If this method is invoked from a shutdown hook
```
Comment (dholmes-ora) [inline src/java.base/share/classes/java/lang/Runtime.java:90]: If they block indefinitely then it is implicit that nothing happens with their exit status. I think you are trying to say too much.

File: src/java.base/share/classes/java/lang/Runtime.java (hunk 3 above)
Comment (rjernst) [reply src/java.base/share/classes/java/lang/Runtime.java:88]: Aha! I think I misread the comment on Shutdown.shutdown and had it stuck in my head that signals would exit through that method. Thanks for the explanation, it makes a lot more sense now.

> All we have to do here is describe how multiple calls to exit() behave.

The thing that is still missing to me is that an application developer may have a single call to System.exit, yet the status code they pass to it may not be the one the jvm exits with.

File: src/java.base/share/classes/java/lang/Runtime.java (hunk 2 above)
Comment (rjernst) [reply src/java.base/share/classes/java/lang/Runtime.java]: I like the new suggested wording, but I would slightly tweak it. As Alan mentioned in another comment the first paragraph makes clear the method never returns normally. How does the following sound?

> If this method is invoked after shutdown hooks have already been started, the supplied status code will be ignored. If this method is invoked from a shutdown hook the system will deadlock.

Comment (dholmes-ora) [reply src/java.base/share/classes/java/lang/Runtime.java]: > gives the impression that an invocation of this method WILL terminate the VM with the given status code - which is not actually true, given the potential for signals.

The subtle distinction being calling Runtime.exit versus Shutdown.exit. The actual serialization takes places in Shutdown.exit and serializes with signals too. But discussion of signals has no place here - I'm not even sure we document those ways of terminating the VM anywhere?

File: src/java.base/share/classes/java/lang/Runtime.java (hunk 3 above)
Comment (dholmes-ora) [reply src/java.base/share/classes/java/lang/Runtime.java:88]: Yes and the problem is? If the user kills the VM by other means; or the termination of the last non-daemon thread races with a call to System.exit then the developer doesn't really have to right to expect anything in relation to that one call.

File: src/java.base/share/classes/java/lang/Runtime.java (hunk 2 above)
Comment (ChrisHegarty) [reply src/java.base/share/classes/java/lang/Runtime.java]: @dholmes-ora Agreed - this paragraph should not explicitly mention signals or any _other_ means of termination.  

There is some shutdown hook specific verbiage relating to termination in `Runtime::addShutdownHook` :

```
In rare circumstances the virtual machine may abort, that is, stop running without shutting
down cleanly. This occurs when the virtual machine is terminated externally, for example
with the SIGKILL signal on Unix or the TerminateProcess call on Microsoft Windows. The
virtual machine may also abort if a native method goes awry by, for example, corrupting
internal data structures or attempting to access nonexistent memory. If the virtual machine
aborts then no guarantee can be made about whether or not any shutdown hooks will be run.
```

... but we do not need to go there, and it's not really enough to leverage anyway. 

>The subtle distinction being calling Runtime.exit versus Shutdown.exit. The actual serialization takes places in Shutdown.exit and serializes with signals too.

Exactly. On further reading of the proposed "Invocations of this method are serialized such that only one invocation will actually proceed with the shutdown sequence and terminate the VM with the given status code.", I think that it is fine (given its position and context in Runtime::exit). My reason for suggesting an alternative ("If this method is invoked after shutdown hooks have already been started, it will block indefinitely") was mainly to try to find something a little simpler while retaining correctness.  Given the discussion so far, either is fine with me.

**Comment by rjernst** [comment]
I appreciate all the feedback and the many opinions expressed here! This has been a learning exercise for me in finding a balance between implementation and specification. While I still think mentioning the possibility of signals is beneficial to a developer trying to understand that the passed status code could be ignored, the text suggested by @dholmes-ora is better than was previously there, so I have updated this PR with that. See [fccd85b](https://github.com/openjdk/jdk/pull/9351/commits/fccd85ba106ff651c00479446ac3207ed60698e8).

**Review by ChrisHegarty — APPROVED** [review]

File: src/java.base/share/classes/java/lang/Runtime.java (hunk 2 above)
[the remark this answers is not in the document]
Comment (kimbarrett) [reply src/java.base/share/classes/java/lang/Runtime.java]: You folks are correct.  I hadn't noticed the ApplicationShutdownHooks thing.  Sorry for the noise.

**Comment by kimbarrett** [comment]
> > Let's say we've run shutdown so run all the hooks but not halted. Then someone calls exit(0). That seems to suggest the call will block indefinitely, which is neither desirable nor what was actually implemented.
> 
> If the hook threads do not halt then the exiting thread (which holds the lock) blocks forever in the join(). Any other call to exit blocks trying to acquire the lock. That has always been the way this works - if hook threads don't terminate then the VM doesn't (unless someone calls halt() directly). That is one of the things the window you suggested be closed, allowed - a call to exit(non-zero) could force a call to halt().

I think that isn't true.  If a hook doesn't terminate then VM.shutdown doesn't get called, so the window never gets cracked opened to call halt directly.

**Comment by ChrisHegarty** [comment]
@dholmes-ora @AlanBateman @kimbarrett The following [CSR](https://bugs.openjdk.org/browse/JDK-8289824) has been filed to cover the specification related changes in this PR - please review as appropriate.

**Review by dholmes-ora — APPROVED** [review]
Thanks for the updates

**Comment by dholmes-ora** [comment]
> I think that isn't true. If a hook doesn't terminate then VM.shutdown doesn't get called, so the window never gets cracked opened to call halt directly.

@kimbarrett  Any thread can call halt() directly while the attempted shutdown is in progress.

**Comment by kimbarrett** [comment]
> > I think that isn't true. If a hook doesn't terminate then VM.shutdown doesn't get called, so the window never gets cracked opened to call halt directly.
> 
> @kimbarrett Any thread can call halt() directly while the attempted shutdown is in progress.

The "that" I was referring to / disagreeing with is about the window for
exit(non-zero), which I think won't ever open if any hook doesn't terminate.

**Review by kimbarrett — APPROVED** [review]
